Compiler error for inner function call with char argument

**Minimal example**: Compile the following code with Clang++ with the clad plugin:

```c++
#include "clad/Differentiator/Differentiator.h"

void g(char c){}

double f(double x) {
  g('a');
  return x;
}

int main() {
  auto f_dx = clad::differentiate(f, "x");
  f_dx.execute(3);
}
```

**Expected behavior**: Code compiles.

**Observed behavior**: Compiler error:

```
charcall.cpp:6:5: warning: attempted to differentiate unsupported statement, no changes applied
  g('a');
    ^
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
...
clang: error: unable to execute command: Segmentation fault (core dumped)
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Ubuntu clang version 11.1.0-6
...
```

**Further information**: The error goes away when the char literal is replaced by an integer literal. The problem also shows up when a string literal is provided for a `char*` argument (which is the setup where I stumbled across this bug, but the above case shows that it's not related to pointers). Many thanks to @vgvassilev and @parth-07 for your quick responses to my earlier bug reports. I'm trying to apply Clad to a particle physics simulation code with as few changes to that code as possible, which seems to break a few things...

**Setup**: Clad built from the GitHub master branch at commit 1bdd261 on Ubuntu 22.04 with Clang 11.

--- a/include/clad/Differentiator/BaseForwardModeVisitor.h
+++ b/include/clad/Differentiator/BaseForwardModeVisitor.h
@@ -56,6 +56,8 @@ public:
   StmtDiff VisitCompoundStmt(const clang::CompoundStmt* CS);
   StmtDiff VisitConditionalOperator(const clang::ConditionalOperator* CO);
   StmtDiff VisitCXXBoolLiteralExpr(const clang::CXXBoolLiteralExpr* BL);
+  StmtDiff VisitCharacterLiteral(const clang::CharacterLiteral* CL);
+  StmtDiff VisitStringLiteral(const clang::StringLiteral* SL);
   StmtDiff VisitCXXDefaultArgExpr(const clang::CXXDefaultArgExpr* DE);
   StmtDiff VisitDeclRefExpr(const clang::DeclRefExpr* DRE);
   StmtDiff VisitDeclStmt(const clang::DeclStmt* DS);

--- a/include/clad/Differentiator/ReverseModeVisitor.h
+++ b/include/clad/Differentiator/ReverseModeVisitor.h
@@ -347,6 +347,8 @@ namespace clad {
     virtual StmtDiff VisitCompoundStmt(const clang::CompoundStmt* CS);
     StmtDiff VisitConditionalOperator(const clang::ConditionalOperator* CO);
     StmtDiff VisitCXXBoolLiteralExpr(const clang::CXXBoolLiteralExpr* BL);
+    StmtDiff VisitCharacterLiteral(const clang::CharacterLiteral* CL);
+    StmtDiff VisitStringLiteral(const clang::StringLiteral* SL);
     StmtDiff VisitCXXDefaultArgExpr(const clang::CXXDefaultArgExpr* DE);
     virtual StmtDiff VisitDeclRefExpr(const clang::DeclRefExpr* DRE);
     StmtDiff VisitDeclStmt(const clang::DeclStmt* DS);

--- a/include/clad/Differentiator/StmtClone.h
+++ b/include/clad/Differentiator/StmtClone.h
@@ -121,11 +121,16 @@ namespace utils {
     DECLARE_CLONE_FN(MaterializeTemporaryExpr)
     DECLARE_CLONE_FN(PseudoObjectExpr)
     DECLARE_CLONE_FN(SubstNonTypeTemplateParmExpr)
+    DECLARE_CLONE_FN(CXXScalarValueInitExpr)
     // `ConstantExpr` node is only available after clang 7.
     #if CLANG_VERSION_MAJOR > 7
     DECLARE_CLONE_FN(ConstantExpr)
     #endif
-    
+#if CLANG_VERSION_MAJOR > 8
+    // `ValueStmt` node is only available after clang 8.
+    DECLARE_CLONE_FN(ValueStmt)
+#endif
+
     clang::Stmt* VisitStmt(clang::Stmt*);
   };
 

--- a/lib/Differentiator/BaseForwardModeVisitor.cpp
+++ b/lib/Differentiator/BaseForwardModeVisitor.cpp
@@ -935,14 +935,14 @@ StmtDiff BaseForwardModeVisitor::VisitDeclRefExpr(const DeclRefExpr* DRE) {
 StmtDiff BaseForwardModeVisitor::VisitIntegerLiteral(const IntegerLiteral* IL) {
   QualType T = IL->getType();
   llvm::APInt zero(m_Context.getIntWidth(T), /*value*/ 0);
-  auto constant0 = IntegerLiteral::Create(m_Context, zero, T, noLoc);
+  auto* constant0 = IntegerLiteral::Create(m_Context, zero, T, noLoc);
   return StmtDiff(Clone(IL), constant0);
 }
 
 StmtDiff
 BaseForwardModeVisitor::VisitFloatingLiteral(const FloatingLiteral* FL) {
   llvm::APFloat zero = llvm::APFloat::getZero(FL->getSemantics());
-  auto constant0 =
+  auto* constant0 =
       FloatingLiteral::Create(m_Context, zero, true, FL->getType(), noLoc);
   return StmtDiff(Clone(FL), constant0);
 }
@@ -1108,7 +1108,7 @@ StmtDiff BaseForwardModeVisitor::VisitCallExpr(const CallExpr* CE) {
         // subexpression.
         if (const auto* MTE = dyn_cast<MaterializeTemporaryExpr>(arg))
           arg = clad_compat::GetSubExpr(MTE);
-        if (!isa<FloatingLiteral>(arg) && !isa<IntegerLiteral>(arg)) {
+        if (!arg->isEvaluatable(m_Context)) {
           allArgsAreConstantLiterals = false;
           break;
         }
@@ -1455,9 +1455,24 @@ BaseForwardModeVisitor::VisitCXXDefaultArgExpr(const CXXDefaultArgExpr* DE) {
 StmtDiff
 BaseForwardModeVisitor::VisitCXXBoolLiteralExpr(const CXXBoolLiteralExpr* BL) {
   llvm::APInt zero(m_Context.getIntWidth(m_Context.IntTy), /*value*/ 0);
-  auto constant0 =
+  auto* constant0 =
       IntegerLiteral::Create(m_Context, zero, m_Context.IntTy, noLoc);
   return StmtDiff(Clone(BL), constant0);
+}
+
+StmtDiff
+BaseForwardModeVisitor::VisitCharacterLiteral(const CharacterLiteral* CL) {
+  llvm::APInt zero(m_Context.getIntWidth(m_Context.IntTy), /*value*/ 0);
+  auto* constant0 =
+      IntegerLiteral::Create(m_Context, zero, m_Context.IntTy, noLoc);
+  return StmtDiff(Clone(CL), constant0);
+}
+
+StmtDiff BaseForwardModeVisitor::VisitStringLiteral(const StringLiteral* SL) {
+  llvm::APInt zero(m_Context.getIntWidth(m_Context.IntTy), /*value*/ 0);
+  auto* constant0 =
+      IntegerLiteral::Create(m_Context, zero, m_Context.IntTy, noLoc);
+  return StmtDiff(Clone(SL), constant0);
 }
 
 StmtDiff BaseForwardModeVisitor::VisitWhileStmt(const WhileStmt* WS) {
@@ -1926,15 +1941,16 @@ StmtDiff BaseForwardModeVisitor::VisitCXXStaticCastExpr(
 StmtDiff BaseForwardModeVisitor::VisitCXXFunctionalCastExpr(
     const clang::CXXFunctionalCastExpr* FCE) {
   StmtDiff castExprDiff = Visit(FCE->getSubExpr());
+  SourceLocation fakeLoc = utils::GetValidSLoc(m_Sema);
   Expr* clonedFCE = m_Sema
                         .BuildCXXFunctionalCastExpr(
-                            FCE->getTypeInfoAsWritten(), FCE->getType(), noLoc,
-                            castExprDiff.getExpr(), noLoc)
+                            FCE->getTypeInfoAsWritten(), FCE->getType(),
+                            fakeLoc, castExprDiff.getExpr(), fakeLoc)
                         .get();
   Expr* derivedFCE = m_Sema
                          .BuildCXXFunctionalCastExpr(
-                             FCE->getTypeInfoAsWritten(), FCE->getType(), noLoc,
-                             castExprDiff.getExpr_dx(), noLoc)
+                             FCE->getTypeInfoAsWritten(), FCE->getType(),
+                             fakeLoc, castExprDiff.getExpr_dx(), fakeLoc)
                          .get();
   return {clonedFCE, derivedFCE};
 }

--- a/lib/Differentiator/DiffPlanner.cpp
+++ b/lib/Differentiator/DiffPlanner.cpp
@@ -562,10 +562,13 @@ namespace clad {
         // bitmask_opts is a template pack of unsigned integers, so we need to
         // do bitwise or of all the values to get the final value.
         unsigned bitmasked_opts_value = 0;
-        for (auto const& arg :
-             FD->getTemplateSpecializationArgs()->get(0).pack_elements()) {
-          bitmasked_opts_value |= arg.getAsIntegral().getExtValue();
-        }
+        const auto template_arg = FD->getTemplateSpecializationArgs()->get(0);
+        if (template_arg.getKind() == TemplateArgument::Pack)
+          for (const auto& arg :
+               FD->getTemplateSpecializationArgs()->get(0).pack_elements())
+            bitmasked_opts_value |= arg.getAsIntegral().getExtValue();
+        else
+          bitmasked_opts_value = template_arg.getAsIntegral().getExtValue();
         unsigned derivative_order =
             clad::GetDerivativeOrder(bitmasked_opts_value);
         if (derivative_order == 0) {
@@ -602,13 +605,15 @@ namespace clad {
         // bitmask_opts is a template pack of unsigned integers, so we need to
         // do bitwise or of all the values to get the final value.
         unsigned bitmasked_opts_value = 0;
-        for (auto const& arg :
-             FD->getTemplateSpecializationArgs()->get(0).pack_elements()) {
-          bitmasked_opts_value |= arg.getAsIntegral().getExtValue();
-        }
-        if (clad::HasOption(bitmasked_opts_value, clad::opts::use_enzyme)) {
+        const auto template_arg = FD->getTemplateSpecializationArgs()->get(0);
+        if (template_arg.getKind() == TemplateArgument::Pack)
+          for (const auto& arg :
+               FD->getTemplateSpecializationArgs()->get(0).pack_elements())
+            bitmasked_opts_value |= arg.getAsIntegral().getExtValue();
+        else
+          bitmasked_opts_value = template_arg.getAsIntegral().getExtValue();
+        if (clad::HasOption(bitmasked_opts_value, clad::opts::use_enzyme))
           request.use_enzyme = true;
-        }
         // reverse vector mode is not yet supported.
         if (clad::HasOption(bitmasked_opts_value, clad::opts::vector_mode)) {
           utils::EmitDiag(m_Sema, DiagnosticsEngine::Error, endLoc,

--- a/lib/Differentiator/ReverseModeVisitor.cpp
+++ b/lib/Differentiator/ReverseModeVisitor.cpp
@@ -1147,6 +1147,17 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
     return Clone(BL);
   }
 
+  StmtDiff
+  ReverseModeVisitor::VisitCharacterLiteral(const CharacterLiteral* CL) {
+    return Clone(CL);
+  }
+
+  StmtDiff ReverseModeVisitor::VisitStringLiteral(const StringLiteral* SL) {
+    return StmtDiff(Clone(SL), StringLiteral::Create(
+                                   m_Context, "", SL->getKind(), SL->isPascal(),
+                                   SL->getType(), utils::GetValidSLoc(m_Sema)));
+  }
+
   StmtDiff ReverseModeVisitor::VisitReturnStmt(const ReturnStmt* RS) {
     // Initially, df/df = 1.
     const Expr* value = RS->getRetValue();
@@ -1386,7 +1397,7 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
         // subexpression.
         if (const auto* MTE = dyn_cast<MaterializeTemporaryExpr>(arg))
           arg = clad_compat::GetSubExpr(MTE)->IgnoreImpCasts();
-        if (!isa<FloatingLiteral>(arg) && !isa<IntegerLiteral>(arg)) {
+        if (!arg->isEvaluatable(m_Context)) {
           allArgsAreConstantLiterals = false;
           break;
         }

--- a/lib/Differentiator/StmtClone.cpp
+++ b/lib/Differentiator/StmtClone.cpp
@@ -83,7 +83,9 @@ Stmt* StmtClone::VisitDeclRefExpr(DeclRefExpr *Node) {
       Ctx, Node->getQualifierLoc(), Node->getTemplateKeywordLoc(),
       Node->getDecl(), Node->refersToEnclosingVariableOrCapture(),
       Node->getNameInfo(), CloneType(Node->getType()), Node->getValueKind(),
-      Node->getFoundDecl(), &TAListInfo);
+      Node->getFoundDecl(),
+      &TAListInfo CLAD_COMPAT_CLANG9_MemberExpr_ExtraParams(
+          Node->isNonOdrUse()));
 }
 DEFINE_CREATE_EXPR(IntegerLiteral,
                    (Ctx, Node->getValue(), CloneType(Node->getType()),
@@ -227,6 +229,9 @@ DEFINE_CLONE_EXPR(VAArgExpr,
                    Node->getWrittenTypeInfo(), Node->getRParenLoc(),
                    CloneType(Node->getType()), Node->isMicrosoftABI()))
 DEFINE_CLONE_EXPR(ImplicitValueInitExpr, (CloneType(Node->getType())))
+DEFINE_CLONE_EXPR(CXXScalarValueInitExpr,
+                  (CloneType(Node->getType()), Node->getTypeSourceInfo(),
+                   Node->getRParenLoc()))
 DEFINE_CLONE_EXPR(ExtVectorElementExpr, (Node->getType(), Node->getValueKind(), Clone(Node->getBase()), Node->getAccessor(), Node->getAccessorLoc()))
 DEFINE_CLONE_EXPR(CXXBoolLiteralExpr, (Node->getValue(), Node->getType(), Node->getSourceRange().getBegin()))
 DEFINE_CLONE_EXPR(CXXNullPtrLiteralExpr, (Node->getType(), Node->getSourceRange().getBegin()))
@@ -429,6 +434,10 @@ DEFINE_CLONE_STMT(BreakStmt, (Node->getBreakLoc()))
 DEFINE_CLONE_STMT(CXXCatchStmt, (Node->getCatchLoc(),
                                  CloneDeclOrNull(Node->getExceptionDecl()),
                                  Clone(Node->getHandlerBlock())))
+
+#if CLANG_VERSION_MAJOR > 8
+DEFINE_CLONE_STMT(ValueStmt, (Node->getStmtClass()))
+#endif
 
 Stmt* StmtClone::VisitCXXTryStmt(CXXTryStmt* Node) {
   llvm::SmallVector<Stmt*, 4> CatchStmts(std::max(1u, Node->getNumHandlers()));

--- a/test/FirstDerivative/CodeGenSimple.C
+++ b/test/FirstDerivative/CodeGenSimple.C
@@ -1,4 +1,4 @@
-// RUN: %cladclang %s -I%S/../../include -oCodeGenSimple.out -Xclang -verify 2>&1 | FileCheck %s
+// RUN: %cladclang %s -I%S/../../include -oCodeGenSimple.out 2>&1 | FileCheck %s
 // RUN: ./CodeGenSimple.out | FileCheck -check-prefix=CHECK-EXEC %s
 
 //CHECK-NOT: {{.*error|warning|note:.*}}
@@ -7,7 +7,7 @@
 extern "C" int printf(const char* fmt, ...);
 
 int f_1(int x) {
-   printf("I am being run!\n"); //expected-warning{{attempted to differentiate unsupported statement, no changes applied}} //expected-warning{{function 'printf' was not differentiated because clad failed to differentiate it and no suitable overload was found in namespace 'custom_derivatives', and function may not be eligible for numerical differentiation.}}
+   printf("I am being run!\n");
    return x * x;
 }
 // CHECK: int f_1_darg0(int x) {

--- a/test/FirstDerivative/FunctionCallsWithResults.C
+++ b/test/FirstDerivative/FunctionCallsWithResults.C
@@ -170,10 +170,6 @@ double nonRealParamFn(const char* a, const char* b = nullptr) {
   return 1;
 }
 
-// CHECK: clad::ValueAndPushforward<double, double> nonRealParamFn_pushforward(const char *a, const char *b, const char *_d_a, const char *_d_b) {
-// CHECK-NEXT:     return {1, 0};
-// CHECK-NEXT: }
-
 double fn4(double i, double j) {
   double res = nonRealParamFn(0, 0);
   res += i;
@@ -183,9 +179,8 @@ double fn4(double i, double j) {
 // CHECK: double fn4_darg0(double i, double j) {
 // CHECK-NEXT:     double _d_i = 1;
 // CHECK-NEXT:     double _d_j = 0;
-// CHECK-NEXT:     clad::ValueAndPushforward<double, double> _t0 = nonRealParamFn_pushforward(0, 0, 0, 0);
-// CHECK-NEXT:     double _d_res = _t0.pushforward;
-// CHECK-NEXT:     double res = _t0.value;
+// CHECK-NEXT:     double _d_res = 0;
+// CHECK-NEXT:     double res = nonRealParamFn(0, 0);
 // CHECK-NEXT:     _d_res += _d_i;
 // CHECK-NEXT:     res += i;
 // CHECK-NEXT:     return _d_res;
@@ -294,10 +289,21 @@ double sum(double* arr, int n) {
 // CHECK-NEXT:     return {val, _d_val};
 // CHECK-NEXT: }
 
+double check_and_return(double x, char c) {
+  if (c == 'a')
+    return x;
+  return 1;
+}
+// CHECK: clad::ValueAndPushforward<double, double> check_and_return_pushforward(double x, char c, double _d_x, char _d_c) {
+// CHECK-NEXT:   if (c == 'a')
+// CHECK-NEXT:     return {x, _d_x};
+// CHECK-NEXT:   return {1, 0};
+// CHECK-NEXT: }
+
 double fn8(double i, double j) {
   double arr[5] = {};
   modifyArr(arr, 5, i*j);
-  return sum(arr, 5) * std::tanh(1.0);
+  return check_and_return(sum(arr, 5), 'a') * std::tanh(1.0);
 }
 
 // CHECK: double fn8_darg0(double i, double j) {
@@ -307,9 +313,10 @@ double fn8(double i, double j) {
 // CHECK-NEXT:     double arr[5] = {};
 // CHECK-NEXT:     modifyArr_pushforward(arr, 5, i * j, _d_arr, 0, _d_i * j + i * _d_j);
 // CHECK-NEXT:     clad::ValueAndPushforward<double, double> _t0 = sum_pushforward(arr, 5, _d_arr, 0);
-// CHECK-NEXT:     double &_t1 = _t0.value; 
-// CHECK-NEXT:     double _t2 = std::tanh(1.); 
-// CHECK-NEXT:     return _t0.pushforward * _t2 + _t1 * 0; 
+// CHECK-NEXT:     clad::ValueAndPushforward<double, double> _t1 = check_and_return_pushforward(_t0.value, 'a', _t0.pushforward, 0);
+// CHECK-NEXT:     double &_t2 = _t1.value;
+// CHECK-NEXT:     double _t3 = std::tanh(1.);
+// CHECK-NEXT:     return _t1.pushforward * _t3 + _t2 * 0;
 // CHECK-NEXT: }
 
 float test_1_darg0(float x);

--- a/test/Gradient/FunctionCalls.C
+++ b/test/Gradient/FunctionCalls.C
@@ -1,9 +1,9 @@
-// RUN: %cladnumdiffclang -std=c++17 %s  -I%S/../../include -oFunctionCalls.out 2>&1 | FileCheck %s
+// RUN: %cladnumdiffclang -std=c++17 -Wno-writable-strings %s  -I%S/../../include -oFunctionCalls.out 2>&1 | FileCheck %s
 // RUN: ./FunctionCalls.out | FileCheck -check-prefix=CHECK-EXEC %s
-// RUN: %cladnumdiffclang -Xclang -plugin-arg-clad -Xclang -enable-tbr -std=c++17 %s  -I%S/../../include -oFunctionCalls.out
+// RUN: %cladnumdiffclang -Xclang -plugin-arg-clad -Xclang -enable-tbr -std=c++17 -Wno-writable-strings %s  -I%S/../../include -oFunctionCalls.out
 // RUN: ./FunctionCalls.out | FileCheck -check-prefix=CHECK-EXEC %s
 
-//CHECK-NOT: {{.*error|warning|note:.*}}
+// CHECK-NOT: {{.*error|warning|note:.*}}
 
 #include "clad/Differentiator/Differentiator.h"
 
@@ -456,20 +456,47 @@ double fn7(double i, double j) {
 // CHECK-NEXT:     }
 // CHECK-NEXT: }
 
+double check_and_return(double x, char c, const char* s) {
+  if (c == 'a' && s[0] == 'a')
+    return x;
+  return 1;
+}
+// CHECK: void check_and_return_pullback(double x, char c, const char *s, double _d_y, clad::array_ref<double> _d_x, clad::array_ref<char> _d_c, clad::array_ref<char> _d_s) {
+// CHECK-NEXT:     bool _cond0;
+// CHECK-NEXT:     _cond0 = c == 'a' && s[0] == 'a';
+// CHECK-NEXT:     if (_cond0)
+// CHECK-NEXT:         goto _label0;
+// CHECK-NEXT:     goto _label1;
+// CHECK-NEXT:   _label1:
+// CHECK-NEXT:     ;
+// CHECK-NEXT:     if (_cond0)
+// CHECK-NEXT:       _label0:
+// CHECK-NEXT:         * _d_x += _d_y;
+// CHECK-NEXT: }
+
 double fn8(double x, double y) {
-  return x*y*std::tanh(1.0)*std::max(1.0, 2.0);
+  return check_and_return(x, 'a', "aa") * y * std::tanh(1.0) * std::max(1.0, 2.0); // expected-warning {{ISO C++11 does not allow conversion from string literal to 'char *' [-Wwritable-strings]}}
 }
 
 // CHECK: void fn8_grad(double x, double y, clad::array_ref<double> _d_x, clad::array_ref<double> _d_y) {
 // CHECK-NEXT:     double _t0;
 // CHECK-NEXT:     double _t1;
+// CHECK-NEXT:     double _t3;
+// CHECK-NEXT:     _t3 = check_and_return(x, 'a', "aa");
 // CHECK-NEXT:     _t1 = std::tanh(1.);
 // CHECK-NEXT:     _t0 = std::max(1., 2.);
 // CHECK-NEXT:     goto _label0;
 // CHECK-NEXT:   _label0:
 // CHECK-NEXT:     {
-// CHECK-NEXT:         * _d_x += 1 * _t0 * _t1 * y;
-// CHECK-NEXT:         * _d_y += x * 1 * _t0 * _t1;
+// CHECK-NEXT:         double _grad0 = 0.;
+// CHECK-NEXT:         char _grad1 = 0i8;
+// CHECK-NEXT:         clad::array_ref<char> _t2 = {"", 3UL};
+// CHECK-NEXT:         check_and_return_pullback(x, 'a', "aa", 1 * _t0 * _t1 * y, &_grad0, &_grad1, _t2);
+// CHECK-NEXT:         double _r0 = _grad0;
+// CHECK-NEXT:         * _d_x += _r0;
+// CHECK-NEXT:         char _r1 = _grad1;
+// CHECK-NEXT:         clad::array<char> _r2({"", 3UL});
+// CHECK-NEXT:         * _d_y += _t3 * 1 * _t0 * _t1;
 // CHECK-NEXT:     }
 // CHECK-NEXT: }
 

